File: src/vscripts/Teams.ts
import { ShuffledList } from "./Utility";



const team_colors = new Map<number, number[]>([
        [DotaTeam.GOODGUYS, [61, 210, 150]], // teal
        [DotaTeam.BADGUYS,  [243, 201, 9]],  // yellow
        [DotaTeam.CUSTOM_1, [197, 77, 168]], // pink
        [DotaTeam.CUSTOM_2, [255, 108, 0]], // orange
        [DotaTeam.CUSTOM_3, [52, 85, 255]], // blue
        [DotaTeam.CUSTOM_4, [101, 212, 19]], // green
        [DotaTeam.CUSTOM_5, [129, 83, 54]], // brown
        [DotaTeam.CUSTOM_6, [27, 192, 216]], // cyan
        [DotaTeam.CUSTOM_7, [199, 228, 13]], //lime
        [DotaTeam.CUSTOM_8, [140, 42, 244]], //purple
])

export  class TeamsSettings{

    public m_VictoryMessages = new Map<number, string>();
    public m_GatheredShuffledTeams: number[] = []
    public maxPlayersPerValidTeam: number = 1;
    constructor() {
        this.SettingTeams()

        this.m_VictoryMessages.set(DotaTeam.GOODGUYS, "#VictoryMessage_GoodGuys");
        this.m_VictoryMessages.set(DotaTeam.BADGUYS, "#VictoryMessage_BadGuys");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_1, "#VictoryMessage_Custom1");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_2, "#VictoryMessage_Custom2");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_3, "#VictoryMessage_Custom3");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_4, "#VictoryMessage_Custom4");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_5, "#VictoryMessage_Custom5");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_6, "#VictoryMessage_Custom6");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_7, "#VictoryMessage_Custom7");
        this.m_VictoryMessages.set(DotaTeam.CUSTOM_8, "#VictoryMessage_Custom8");
    }

    public SettingTeams(){
        this.HealthBarChange()



        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.GOODGUYS, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.BADGUYS, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_1, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_2, 1);    
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_3, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_4, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_5, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_6, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_7, 1);
        GameRules.SetCustomGameTeamMaxPlayers(DotaTeam.CUSTOM_8, 1);
    }

     // Хуйня какая-то, переделать!(для других карт плохо работет)
    public GatherAndRegisterValidTeams() : number[]{
        
    //print( "GatherValidTeams:" )

        const foundTeams = new Set<number>()// список ключ - id команды
        for(const playerStart of Entities.FindAllByClassname( "info_player_start_dota")){
            foundTeams.add(playerStart.GetTeamNumber())
            //print("FoundTeams adding " + playerStart.GetTeamNumber().toString() + " " + GetTeamName(playerStart.GetTeam()))
        }
        //print(foundTeams.length)

        ///print( "GatherValidTeams - Found spawns for(a total of " +  numTeams  + " teams" )

        const foundTeamsList: number[] = []
        print( this.m_VictoryMessages.get(0), this.m_VictoryMessages.get(1))
        foundTeams.forEach( (v,i,o) => {
            //если команда с таким id существует (team), как признак у нее есть victory message
            if(this.m_VictoryMessages.get(v)!=undefined){
                table.insert( foundTeamsList, v)
               // print( "FoundTeams.ForEach: " + tostring(v) + " " + this.m_VictoryMessages.get(v))
            }

        })
        //print("FoundTeams.length = " + tostring(foundTeamsList.length) )

        /*
        for(const [t, _] of pairs( foundTeams ) ){
            table.insert( foundTeamsList, t )
            print("FoundTeamList creation: Team Adding  " + t.toString() + (this.m_VictoryMessages[t]!=undefined?this.m_VictoryMessages[t]:"victory message undefined"))
        }
        */
        let numTeams = foundTeamsList.length
        //print("FoundTeamList.length = " + foundTeamsList.length)

        if (numTeams == 0) {
           // print( "GatherValidTeams - NO team spawns detected, defaulting to GOOD/BAD" )
            table.insert( foundTeamsList, DotaTeam.GOODGUYS )
            table.insert( foundTeamsList, DotaTeam.BADGUYS )
            numTeams = 2
        }
        this.maxPlayersPerValidTeam = math.floor( 10 / numTeams )
        //print("GetMapName(): " + GetMapName())
        switch(GetMapName()){
            case "forest_solo":{
                this.maxPlayersPerValidTeam = 1
                break;
            }
            case "desert_duo":{
                this.maxPlayersPerValidTeam = 2
                break;
            }
            case "temple_quartet":{
                this.maxPlayersPerValidTeam = 4
                break;
            }
            case "desert_quintet":{
                this.maxPlayersPerValidTeam = 5
                break;
            }
            default:{
                this.maxPlayersPerValidTeam = 3
                break;
            }
        }

       // print("maxPlayersPerValidTeam = " + tostring(maxPlayersPerValidTeam) + " numteams = " + numTeams)
        this.m_GatheredShuffledTeams = ShuffledList( foundTeamsList )

        /*
        print( "Final shuffled team list:" )
        for (const [_, team] of pairs( this.m_GatheredShuffledTeams )) {
            print( " - " + team + " "  + GetTeamName( team ))
        } 
        */

        //print( "Setting up teams:" )

        for(const team  of $range(0, (DOTA_TEAM_COUNT-1) )){
            GameRules.SetCustomGameTeamMaxPlayers( team, 0  )
        }
        for(let i = 0; i < numTeams; i++) {
            //print( " - " + this.m_GatheredShuffledTeams[i] + "  " + GetTeamName( this.m_GatheredShuffledTeams[i] ) + " max players = " + tostring(maxPlayersPerValidTeam) )
            GameRules.SetCustomGameTeamMaxPlayers( this.m_GatheredShuffledTeams[i], this.maxPlayersPerValidTeam )
        }
        return this.m_GatheredShuffledTeams
    }



    public HealthBarChange() : void {
        for(let el of team_colors){
            const team = el[0]
            const color = el[1]
            SetTeamCustomHealthbarColor(team, color[0], color[1], color[2]);
        }

    }

    //Нахуя надо ?* 
    public UpdatePlayerColor( nPlayerID : PlayerID){
        const teamID = PlayerResource.GetTeam( nPlayerID )
        const color = ColorForTeam( teamID )
        if (!PlayerResource.HasSelectedHero( nPlayerID ) )
            return
        

        const hero = PlayerResource.GetSelectedHeroEntity( nPlayerID )
        if (hero == undefined)
            return
        
        PlayerResource.SetCustomPlayerColor(nPlayerID, color[0], color[1], color[2] )

    }



}
export function ColorForTeam( teamID : number ){
    let color = team_colors.get(teamID)
    if (color == undefined)
        color = [ 255, 255, 255 ] // default to white
    return color
}
export function AssignTeams(m_bFillWithBots : boolean){
    //print( "Assigning teams" )
    const vecTeamValid : boolean[] = []
    const vecTeamNeededPlayers : number[] = []
    for (let nTeam = 0; nTeam < DOTA_TEAM_COUNT; nTeam++) {
        const nMax = GameRules.GetCustomGameTeamMaxPlayers( nTeam )
        if (nMax > 0) {
            //print( "Found team " .. nTeam .. " with max players " .. nMax )
            vecTeamNeededPlayers[ nTeam ] = nMax
            vecTeamValid[ nTeam ] = true
        } else {
            vecTeamValid[ nTeam ] = false
        }
    }

    //loop 1: count up players on each team
    const hPlayers : PlayerID[]= []
    for (let nPlayerID = 0; nPlayerID < DOTA_MAX_TEAM_PLAYERS; nPlayerID++) {
        if (PlayerResource.IsValidPlayerID( nPlayerID )) {
            let nTeam = PlayerResource.GetTeam( nPlayerID )
            if (!vecTeamValid[ nTeam ]) {
                nTeam = PlayerResource.GetCustomTeamAssignment( nPlayerID )
            }
            //print( "Found player " .. nPlayerID .. " on team " .. nTeam )
            if (vecTeamValid[ nTeam ]) {
                vecTeamNeededPlayers[ nTeam ] = vecTeamNeededPlayers[ nTeam ] - 1
            } else {
                hPlayers.push( nPlayerID )
            }
        }
    }

    //loop 2: assign players. For each player who is on an invalid team,
    //find the team that has the highest number of needed players
    //&& assign the player to that team
    hPlayers.forEach( (nPlayerID, _) =>  {
        //print( "Finding team for player " .. nPlayerID )
        let nTeamNumber = -1
        let nHighest = 0
        for (let nTeam = 0; nTeam < DOTA_TEAM_COUNT; nTeam++) {
            if (vecTeamValid[ nTeam ]) {
                const nVal = vecTeamNeededPlayers[ nTeam ]
                if (nVal > nHighest) {
                    //print( "found team " .. nTeam .. " with needed " .. nVal .. " but highest was only " .. nHighest )
                    nHighest = nVal
                    nTeamNumber = nTeam
                }
            }
        }
        if (nTeamNumber > 0) {
            PlayerResource.SetCustomTeamAssignment( nPlayerID , nTeamNumber )
            vecTeamNeededPlayers[ nTeamNumber ] = vecTeamNeededPlayers[ nTeamNumber ] - 1
        }
    })
        
    if (m_bFillWithBots) {
        GameRules.BotPopulate()
    }
}

File: game/scripts/vscripts/Teams.lua
local ____lualib = require("lualib_bundle")
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__Class = ____lualib.__TS__Class
local Set = ____lualib.Set
local __TS__Iterator = ____lualib.__TS__Iterator
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["11"] = 5,["12"] = 1,["13"] = 1,["14"] = 173,["15"] = 174,["16"] = 175,["17"] = 176,["19"] = 177,["20"] = 173,["21"] = 5,["22"] = 5,["23"] = 5,["24"] = 5,["25"] = 5,["26"] = 5,["27"] = 5,["28"] = 5,["29"] = 5,["30"] = 5,["31"] = 5,["32"] = 5,["33"] = 18,["34"] = 18,["35"] = 18,["37"] = 20,["38"] = 21,["39"] = 22,["40"] = 24,["41"] = 26,["42"] = 27,["43"] = 28,["44"] = 29,["45"] = 30,["46"] = 31,["47"] = 32,["48"] = 33,["49"] = 34,["50"] = 35,["51"] = 23,["52"] = 38,["53"] = 39,["54"] = 43,["55"] = 44,["56"] = 45,["57"] = 46,["58"] = 47,["59"] = 48,["60"] = 49,["61"] = 50,["62"] = 51,["63"] = 52,["64"] = 38,["65"] = 56,["66"] = 60,["67"] = 61,["68"] = 62,["70"] = 69,["71"] = 70,["72"] = 70,["73"] = 70,["74"] = 70,["75"] = 71,["76"] = 73,["77"] = 74,["79"] = 71,["80"] = 87,["81"] = 90,["82"] = 92,["83"] = 93,["84"] = 94,["86"] = 96,["88"] = 98,["89"] = 99,["92"] = 100,["96"] = 103,["99"] = 104,["103"] = 107,["106"] = 108,["110"] = 111,["113"] = 112,["119"] = 116,["124"] = 122,["125"] = 133,["126"] = 134,["129"] = 136,["130"] = 136,["131"] = 138,["132"] = 136,["135"] = 140,["136"] = 56,["137"] = 145,["138"] = 146,["139"] = 147,["140"] = 148,["141"] = 149,["143"] = 145,["144"] = 155,["145"] = 156,["146"] = 157,["147"] = 158,["150"] = 162,["151"] = 163,["154"] = 166,["155"] = 155,["156"] = 179,["157"] = 181,["158"] = 182,["160"] = 183,["161"] = 183,["162"] = 184,["163"] = 185,["164"] = 187,["165"] = 188,["167"] = 190,["169"] = 183,["172"] = 195,["174"] = 196,["175"] = 196,["176"] = 197,["177"] = 198,["178"] = 199,["179"] = 200,["181"] = 203,["182"] = 204,["184"] = 206,["187"] = 196,["190"] = 214,["191"] = 214,["192"] = 214,["193"] = 216,["194"] = 217,["196"] = 218,["197"] = 218,["198"] = 219,["199"] = 220,["200"] = 221,["201"] = 223,["202"] = 224,["205"] = 218,["208"] = 228,["209"] = 229,["210"] = 230,["212"] = 214,["213"] = 214,["214"] = 234,["215"] = 235,["217"] = 179});
local ____exports = {}
local team_colors
local ____Utility = require("Utility")
local ShuffledList = ____Utility.ShuffledList
function ____exports.ColorForTeam(self, teamID)
    local color = team_colors:get(teamID)
    if color == nil then
        color = {255, 255, 255}
    end
    return color
end
team_colors = __TS__New(Map, {
    {DOTA_TEAM_GOODGUYS, {61, 210, 150}},
    {DOTA_TEAM_BADGUYS, {243, 201, 9}},
    {DOTA_TEAM_CUSTOM_1, {197, 77, 168}},
    {DOTA_TEAM_CUSTOM_2, {255, 108, 0}},
    {DOTA_TEAM_CUSTOM_3, {52, 85, 255}},
    {DOTA_TEAM_CUSTOM_4, {101, 212, 19}},
    {DOTA_TEAM_CUSTOM_5, {129, 83, 54}},
    {DOTA_TEAM_CUSTOM_6, {27, 192, 216}},
    {DOTA_TEAM_CUSTOM_7, {199, 228, 13}},
    {DOTA_TEAM_CUSTOM_8, {140, 42, 244}}
})
____exports.TeamsSettings = __TS__Class()
local TeamsSettings = ____exports.TeamsSettings
TeamsSettings.name = "TeamsSettings"
function TeamsSettings.prototype.____constructor(self)
    self.m_VictoryMessages = __TS__New(Map)
    self.m_GatheredShuffledTeams = {}
    self.maxPlayersPerValidTeam = 1
    self:SettingTeams()
    self.m_VictoryMessages:set(DOTA_TEAM_GOODGUYS, "#VictoryMessage_GoodGuys")
    self.m_VictoryMessages:set(DOTA_TEAM_BADGUYS, "#VictoryMessage_BadGuys")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_1, "#VictoryMessage_Custom1")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_2, "#VictoryMessage_Custom2")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_3, "#VictoryMessage_Custom3")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_4, "#VictoryMessage_Custom4")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_5, "#VictoryMessage_Custom5")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_6, "#VictoryMessage_Custom6")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_7, "#VictoryMessage_Custom7")
    self.m_VictoryMessages:set(DOTA_TEAM_CUSTOM_8, "#VictoryMessage_Custom8")
end
function TeamsSettings.prototype.SettingTeams(self)
    self:HealthBarChange()
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_GOODGUYS, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_BADGUYS, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_1, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_2, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_3, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_4, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_5, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_6, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_7, 1)
    GameRules:SetCustomGameTeamMaxPlayers(DOTA_TEAM_CUSTOM_8, 1)
end
function TeamsSettings.prototype.GatherAndRegisterValidTeams(self)
    local foundTeams = __TS__New(Set)
    for ____, playerStart in ipairs(Entities:FindAllByClassname("info_player_start_dota")) do
        foundTeams:add(playerStart:GetTeamNumber())
    end
    local foundTeamsList = {}
    print(
        self.m_VictoryMessages:get(0),
        self.m_VictoryMessages:get(1)
    )
    foundTeams:forEach(function(____, v, i, o)
        if self.m_VictoryMessages:get(v) ~= nil then
            table.insert(foundTeamsList, v)
        end
    end)
    local numTeams = #foundTeamsList
    if numTeams == 0 then
        table.insert(foundTeamsList, DOTA_TEAM_GOODGUYS)
        table.insert(foundTeamsList, DOTA_TEAM_BADGUYS)
        numTeams = 2
    end
    self.maxPlayersPerValidTeam = math.floor(10 / numTeams)
    repeat
        local ____switch10 = GetMapName()
        local ____cond10 = ____switch10 == "forest_solo"
        if ____cond10 then
            do
                self.maxPlayersPerValidTeam = 1
                break
            end
        end
        ____cond10 = ____cond10 or ____switch10 == "desert_duo"
        if ____cond10 then
            do
                self.maxPlayersPerValidTeam = 2
                break
            end
        end
        ____cond10 = ____cond10 or ____switch10 == "temple_quartet"
        if ____cond10 then
            do
                self.maxPlayersPerValidTeam = 4
                break
            end
        end
        ____cond10 = ____cond10 or ____switch10 == "desert_quintet"
        if ____cond10 then
            do
                self.maxPlayersPerValidTeam = 5
                break
            end
        end
        do
            do
                self.maxPlayersPerValidTeam = 3
                break
            end
        end
    until true
    self.m_GatheredShuffledTeams = ShuffledList(nil, foundTeamsList)
    for team = 0, DOTA_TEAM_COUNT - 1 do
        GameRules:SetCustomGameTeamMaxPlayers(team, 0)
    end
    do
        local i = 0
        while i < numTeams do
            GameRules:SetCustomGameTeamMaxPlayers(self.m_GatheredShuffledTeams[i + 1], self.maxPlayersPerValidTeam)
            i = i + 1
        end
    end
    return self.m_GatheredShuffledTeams
end
function TeamsSettings.prototype.HealthBarChange(self)
    for ____, el in __TS__Iterator(team_colors) do
        local team = el[1]
        local color = el[2]
        SetTeamCustomHealthbarColor(team, color[1], color[2], color[3])
    end
end
function TeamsSettings.prototype.UpdatePlayerColor(self, nPlayerID)
    local teamID = PlayerResource:GetTeam(nPlayerID)
    local color = ____exports.ColorForTeam(nil, teamID)
    if not PlayerResource:HasSelectedHero(nPlayerID) then
        return
    end
    local hero = PlayerResource:GetSelectedHeroEntity(nPlayerID)
    if hero == nil then
        return
    end
    PlayerResource:SetCustomPlayerColor(nPlayerID, color[1], color[2], color[3])
end
function ____exports.AssignTeams(self, m_bFillWithBots)
    local vecTeamValid = {}
    local vecTeamNeededPlayers = {}
    do
        local nTeam = 0
        while nTeam < DOTA_TEAM_COUNT do
            local nMax = GameRules:GetCustomGameTeamMaxPlayers(nTeam)
            if nMax > 0 then
                vecTeamNeededPlayers[nTeam + 1] = nMax
                vecTeamValid[nTeam + 1] = true
            else
                vecTeamValid[nTeam + 1] = false
            end
            nTeam = nTeam + 1
        end
    end
    local hPlayers = {}
    do
        local nPlayerID = 0
        while nPlayerID < DOTA_MAX_TEAM_PLAYERS do
            if PlayerResource:IsValidPlayerID(nPlayerID) then
                local nTeam = PlayerResource:GetTeam(nPlayerID)
                if not vecTeamValid[nTeam + 1] then
                    nTeam = PlayerResource:GetCustomTeamAssignment(nPlayerID)
                end
                if vecTeamValid[nTeam + 1] then
                    vecTeamNeededPlayers[nTeam + 1] = vecTeamNeededPlayers[nTeam + 1] - 1
                else
                    hPlayers[#hPlayers + 1] = nPlayerID
                end
            end
            nPlayerID = nPlayerID + 1
        end
    end
    __TS__ArrayForEach(
        hPlayers,
        function(____, nPlayerID, _)
            local nTeamNumber = -1
            local nHighest = 0
            do
                local nTeam = 0
                while nTeam < DOTA_TEAM_COUNT do
                    if vecTeamValid[nTeam + 1] then
                        local nVal = vecTeamNeededPlayers[nTeam + 1]
                        if nVal > nHighest then
                            nHighest = nVal
                            nTeamNumber = nTeam
                        end
                    end
                    nTeam = nTeam + 1
                end
            end
            if nTeamNumber > 0 then
                PlayerResource:SetCustomTeamAssignment(nPlayerID, nTeamNumber)
                vecTeamNeededPlayers[nTeamNumber + 1] = vecTeamNeededPlayers[nTeamNumber + 1] - 1
            end
        end
    )
    if m_bFillWithBots then
        GameRules:BotPopulate()
    end
end
return ____exports
